### ISSUE TITLE
File location problem

### ISSUE THREAD

**Issue by PhdShi** [pytorch/pytorch#91693]
### 🐛 Describe the bug

hi，when I compile thirty party collective communication library for pytorch 1.13 supported by NGC images "nvcr.io/nvidia/pytorch:22.11-py3", I find that some head files don't exist, just like the following example.
```fatal error: c10d/Utils.hpp: No such file or directory```
When I tried to find it, I found that the latest version of the NGC image did not have conda installed, and there were no csrc files in “/usr/local/lib/python3.8/dist-packages”. So I want to know, if I need this header file, where should I find it.

### Versions

image: nvcr.io/nvidia/pytorch:22.11-py3
Pytorch version: 1.13
CUDA Version: 11.8

cc @ezyang @seemethere @malfet @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

**Comment by viveksingh-ctrl** [comment]
Can you try `python setup.py clean` ?

**Comment by PhdShi** [comment]
While browsing the Nvidia website, I noticed a change in the image for version 22.11:
```
Starting with the 22.11 PyTorch NGC container, miniforge is removed and all Python packages are installed in the default Python environment. In case you depend on Conda-specific packages, which might not be available on PyPI, we recommend building these packages from source. A workaround is to manually install a Conda package manager, and add the conda path to your PYTHONPATH for example, using export PYTHONPATH="/opt/conda/lib/python3.8/site-packages" if your Conda package manager was installed in /opt/conda.
```
But when I compare the torch in python to the torch in conda, I find that there is no c10d directory in the python environment, which means that I can no longer use the relevant head files to implement my third-party collective communication library. Why does Nvidia delete the C10D directory in the new ngc image?

**Comment by mikaylagawarecki** [comment]
Hi, would you be able to provide a minimal reproduction of this please? Thank you!

**Comment by PhdShi** [comment]
> Hi, would you be able to provide a minimal reproduction of this please? Thank you!

Thx for your reply! The following error occurs when I compile torch-ccl in the ngc image "nvcr.io/nvidia/pytorch:22.11-py3": 
```
In file included from /torch-ccl/src/cpu/cpu_ccl.cpp:40:
/torch-ccl/src/./ProcessGroupCCL.hpp:42:10: fatal error: torch/csrc/distributed/c10d/ProcessGroup.hpp: No such file or directory
   42 | #include <torch/csrc/distributed/c10d/ProcessGroup.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [src/CMakeFiles/oneccl_bindings_for_pytorch.dir/build.make:146: src/CMakeFiles/oneccl_bindings_for_pytorch.dir/cpu/cpu_ccl.cpp.o] Error 1
make[3]: *** Waiting for unfinished jobs....
In file included from /torch-ccl/src/ccl_comm_collector.h:37,
                 from /torch-ccl/src/ccl_comm_collector.cpp:4:
/torch-ccl/src/ProcessGroupCCL.hpp:42:10: fatal error: torch/csrc/distributed/c10d/ProcessGroup.hpp: No such file or directory
   42 | #include <torch/csrc/distributed/c10d/ProcessGroup.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [src/CMakeFiles/oneccl_bindings_for_pytorch.dir/build.make:118: src/CMakeFiles/oneccl_bindings_for_pytorch.dir/ccl_comm_collector.cpp.o] Error 1
In file included from /torch-ccl/src/dispatch_stub.h:34,
                 from /torch-ccl/src/dispatch_stub.cpp:34:
/torch-ccl/src/utils.h:41:10: fatal error: torch/csrc/distributed/c10d/Types.hpp: No such file or directory
   41 | #include <torch/csrc/distributed/c10d/Types.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /torch-ccl/src/utils.cpp:32:
/torch-ccl/src/utils.h:41:10: fatal error: torch/csrc/distributed/c10d/Types.hpp: No such file or directory
   41 | #include <torch/csrc/distributed/c10d/Types.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
compilation terminated.
make[3]: *** [src/CMakeFiles/oneccl_bindings_for_pytorch.dir/build.make:90: src/CMakeFiles/oneccl_bindings_for_pytorch.dir/dispatch_stub.cpp.o] Error 1
make[3]: *** [src/CMakeFiles/oneccl_bindings_for_pytorch.dir/build.make:104: src/CMakeFiles/oneccl_bindings_for_pytorch.dir/utils.cpp.o] Error 1
In file included from /torch-ccl/src/./ccl_comm_collector.h:37,
                 from /torch-ccl/src/ProcessGroupCCL.cpp:36:
/torch-ccl/src/./ProcessGroupCCL.hpp:42:10: fatal error: torch/csrc/distributed/c10d/ProcessGroup.hpp: No such file or directory
   42 | #include <torch/csrc/distributed/c10d/ProcessGroup.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```
But this problem can be solved by recompiling pytorch. This does not seem to be a pytorch issue, but rather a problem with the ngc image deleting the pytorch csrc/c10d header files. I really don't understand why the pytorch provided by ngc image removes the corresponding C++ header file?

**Comment by mikaylagawarecki** [comment]
cc @ngimel @malfet Not quite sure who is the right POC regarding NVIDIA NGC PyTorch image, would you be able to help here?

**Comment by ngimel** [comment]
cc @ptrblck

**Comment by malfet** [comment]
Hmm, this looks like a https://github.com/intel/torch-ccl issue rather than NCG/pytorch one. I.e. https://github.com/intel/torch-ccl/commit/90fb7704b633ce92c216b511c86df488e8a9c3fa assumes that headers are available at the new path, but there are no changes in the installation path, see https://github.com/pytorch/pytorch/pull/85780
@minsii do you know how c10d headers are supposed to be used from 3rd party libraries?

**Comment by ptrblck** [comment]
Thanks for pinging me on this issue.
I think @malfet might be right and it seems the binaries or source builds do not ship with `c10d` in their `include` folder anymore.
Using the pip wheels for `1.12.1` shows:
```
# ls /usr/local/lib/python3.8/dist-packages/torch/include/
ATen  THC  c10  c10d  caffe2  pybind11  torch
# ls /usr/local/lib/python3.8/dist-packages/torch/include/c10d/
FileStore.hpp          ProcessGroupNCCL.hpp        UnixSockUtils.hpp       logger.hpp
GlooDeviceFactory.hpp  ProcessGroupRoundRobin.hpp  Utils.hpp               logging.h
HashStore.hpp          ProcessGroupWrapper.hpp     WinSockUtils.hpp        python_comm_hook.h
NCCLUtils.hpp          PyProcessGroup.hpp          c10d.h                  reducer.hpp
ParamCommsUtils.hpp    Store.hpp                   comm.hpp                reducer_timer.hpp
PrefixStore.hpp        TCPStore.hpp                debug.h                 sequence_num.hpp
ProcessGroup.hpp       TraceUtils.h                default_comm_hooks.hpp  socket.h
ProcessGroupGloo.hpp   Types.hpp                   error.h
ProcessGroupMPI.hpp    UCCForNCCL.hpp              exception.h
```
while `1.13.0` shows:
```
# ls /usr/local/lib/python3.8/dist-packages/torch/include/
ATen  THC  c10  caffe2  pybind11  torch
```
so I don't think our build process for the container is at fault.

**Comment by PhdShi** [comment]
> Thanks for pinging me on this issue. I think @malfet might be right and it seems the binaries or source builds do not ship with `c10d` in their `include` folder anymore. Using the pip wheels for `1.12.1` shows:
> 
> ```
> # ls /usr/local/lib/python3.8/dist-packages/torch/include/
> ATen  THC  c10  c10d  caffe2  pybind11  torch
> # ls /usr/local/lib/python3.8/dist-packages/torch/include/c10d/
> FileStore.hpp          ProcessGroupNCCL.hpp        UnixSockUtils.hpp       logger.hpp
> GlooDeviceFactory.hpp  ProcessGroupRoundRobin.hpp  Utils.hpp               logging.h
> HashStore.hpp          ProcessGroupWrapper.hpp     WinSockUtils.hpp        python_comm_hook.h
> NCCLUtils.hpp          PyProcessGroup.hpp          c10d.h                  reducer.hpp
> ParamCommsUtils.hpp    Store.hpp                   comm.hpp                reducer_timer.hpp
> PrefixStore.hpp        TCPStore.hpp                debug.h                 sequence_num.hpp
> ProcessGroup.hpp       TraceUtils.h                default_comm_hooks.hpp  socket.h
> ProcessGroupGloo.hpp   Types.hpp                   error.h
> ProcessGroupMPI.hpp    UCCForNCCL.hpp              exception.h
> ```
> 
> while `1.13.0` shows:
> 
> ```
> # ls /usr/local/lib/python3.8/dist-packages/torch/include/
> ATen  THC  c10  caffe2  pybind11  torch
> ```
> 
> so I don't think our build process for the container is at fault.

Thx for your reply! This problem has been solved by recompiling pytorch 1.13.0 and ```c10d``` head files I need is in right directory. Also I checked the December image ```nvcr.io/nvidia/pytorch:22.12-py3``` and found that it contains c10d head files. This suggests that this is just an image problem in NVIDIA NGC PyTorch image ```nvcr.io/nvidia/pytorch:22.11-py3```

**Comment by minsii** [comment]
> Hmm, this looks like a https://github.com/intel/torch-ccl issue rather than NCG/pytorch one. I.e. [intel/torch-ccl@90fb770](https://github.com/intel/torch-ccl/commit/90fb7704b633ce92c216b511c86df488e8a9c3fa) assumes that headers are available at the new path, but there are no changes in the installation path, see #85780 @minsii do you know how c10d headers are supposed to be used from 3rd party libraries?

Looks like the specific issue has already been figured out. In case you want to double check the latest usage of c10d headers in 3rd party library, please refer to this tutorial (with Pytorch 1.13): https://pytorch.org/tutorials/intermediate/process_group_cpp_extension_tutorial.html
